### ISSUE TITLE
[Obsolete(error: true)] ignored in code marked `[Obsolete]`

### ISSUE THREAD

**Issue by richard-fine** [dotnet/roslyn#80961]
**Version Used**: Latest (.NET 9)

**Steps to Reproduce**:

```
using System;

[Obsolete("You must not use this", error: true)]
void ObsoleteAsErrorMethod() {
    Console.WriteLine("This should never be invoked because use of the symbol is now an error");
}

[Obsolete("Using this is not recommended", error: false)]
void ObsoleteAsWarningMethod() {
     ObsoleteAsErrorMethod();
}

ObsoleteAsWarningMethod();
```

**Expected Behavior**:

The call to `ObsoleteAsErrorMethod()` should produce a compiler error.

**Actual Behavior**:

The call to `ObsoleteAsWarningMethod()` produces a compiler _warning_, but the subsequent call to `ObsoleteAsErrorMethod()` produces no diagnostics.

**Comment by CyrusNajmabadi** [comment]
This is by design. The existing code needs to be able to call the obsolete code so that the program can continue running.

**Comment by richard-fine** [comment]
If the code were marked only with `[Obsolete(error: false)]` then I agree it makes sense to suppress any warnings when one piece of obsolete code is calling another piece of obsolete code.

However, if I have marked the code with `[Obsolete(error: true)]`, am I not expressly trying to indicate that I don't _want_ the program to continue running until the use of the obsolete API has been removed?

**Comment by CyrusNajmabadi** [comment]
> However, if I have marked the code with [Obsolete(error: true)], am I not expressly trying to indicate that I don't want the program to continue running until the use of the obsolete API has been removed?

No. That has never been the semantics.  It's indicating that non obsolete could should be pushed away from it. But of course there may need to be some obsolete code that still exists and continues to use it.

**Comment by richard-fine** [comment]
> It's indicating that non obsolete could should be pushed away from it. But of course there may need to be some obsolete code that still exists and continues to use it.

But isn't that what `error: false` is for? If you use the attribute with `error: false`, you're already pushing away users with a compiler warning, IDE strikeouts, etc. I always thought the point of `error: true` was to be able to provide a 'tombstone' for code that had been effectively deleted, so that anyone attempting to use it could get a more informative error than just a failure to find the symbol.

[Section 23.5.4](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/attributes#2354-the-obsolete-attribute) doesn't mention anything about this behaviour. If it's by design, perhaps the spec should be updated to mention it?

**Comment by CyrusNajmabadi** [comment]
> But isn't that what error: false is for? If you use the attribute with error: false, you're already pushing away users with a compiler warning, 

You may still need to call it from your code, which is not itself exposed as obsolete-error.

**Comment by jjonescz** [comment]
> [Section 23.5.4](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/attributes#2354-the-obsolete-attribute) doesn't mention anything about this behaviour. If it's by design, perhaps the spec should be updated to mention it?

That would need an issue at https://github.com/dotnet/csharpstandard.

Otherwise I agree this feels like by design. If it were to be changed, it would need to be discussed first over at https://github.com/dotnet/csharplang.

Thanks.
